### DISCUSSION TITLE
Include estimate of bandwidth leaving a data centre

### DISCUSSION THREAD

**Discussion by pwithnall** [cloud-carbon-footprint/cloud-carbon-footprint#1045] [Feature suggestions]
## Feature Suggestion

Currently the [methodology](https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/blob/trunk/microsite/docs/Methodology.md#user-content-scope) says:
> We also have chosen to ignore traffic that leaves a data center to provide end-users with services, because this traffic is usually handled by CDN providers

In my use case, that traffic goes straight to end users, and not to a CDN. Would it be possible to include the end-user traffic using a configuration option please? The option could be disabled (i.e. ignore traffic to end-users) by default.

It would be even more helpful if the bar for (say) S3 in the ‘Emissions Breakdown’ could be split into several stacked sub-bars, one for storage emissions, one for bandwidth emissions to other data centres, one for bandwidth emissions to end users, etc. That would give users of CCF a bit more guidance about which where the main contributions to emissions for each service come from. But that could be a separate feature request (and I’m happy to split it out into a separate issue if that would be more helpful). Thanks.

**Comment by ericksod** [comment]
Hi @pwithnall! Sorry for the late reply. We're going to discuss both of these in further detail. We've been thinking about additional breakdowns to provide more info but need to continue our analysis and are grateful for your input. We'll get back to you once we've discussed these points.

**Comment by dtoakley** [comment]
Hey @pwithnall -- re: networking usage from cloud providers to end users, this is definitely something that should be possible, but the main blocker from us doing this so far is that  whatever co-efficient we use for kilowatt hours per Gigabyte transferred is dependent on the type of end user device (phone, tablet, laptop, etc) and the version (old, new, etc). Given that we don't know the type of user devices from the cloud provider APIs, it's a little challenging to settle on a co-efficient. So it's less about there being a CDN in your stack or not, and more about knowing the types of devices end users are using to request your data. That said, it's definitely possible to implement - we'd just need to do some research to settle on a reasonable coefficient that would make sense for the majority of users of the software as a default.     

Re: your 2nd request, it would be good to separate the request for additional breakdowns of the emissions data based on the type of usage considered when estimating energy/CO2e, if you're able to. If you can include some example use cases and possibly a mockup/image of the stacked bar chart you mention, even better! Thank you!

**Comment by pwithnall** [comment]
> Hey @pwithnall -- re: networking usage from cloud providers to end users, this is definitely something that should be possible, but the main blocker from us doing this so far is that whatever co-efficient we use for kilowatt hours per Gigabyte transferred is dependent on the type of end user device (phone, tablet, laptop, etc) and the version (old, new, etc).

Are you planning on including the power consumption of the end user device in the CCF figures, then, or are you thinking of just using this as a proxy for the type of network infrastructure the traffic goes over (fibre, ethernet, satellite, Wi-Fi, 4G, etc.) which, aiui, has a significant impact on the network emissions?

I suggest that including the power consumption of the end user device itself could easily lead to double counting of emissions (depending on where people are drawing their system boundaries) and, as you’ve said, it introduces a significant number of additional dimensions to model.

> Given that we don't know the type of user devices from the cloud provider APIs, it's a little challenging to settle on a co-efficient. So it's less about there being a CDN in your stack or not, and more about knowing the types of devices end users are using to request your data. That said, it's definitely possible to implement - we'd just need to do some research to settle on a reasonable coefficient that would make sense for the majority of users of the software as a default.

That makes sense. I assume you’d pick a general-purpose default, but allow the coefficients to be customised in the configuration, since site owners are likely to know the rough breakdown of types of network infrastructure their users use, either through having an idea of their target audience, or through web analytics?

> Re: your 2nd request, it would be good to separate the request for additional breakdowns of the emissions data based on the type of usage considered when estimating energy/CO2e, if you're able to. If you can include some example use cases and possibly a mockup/image of the stacked bar chart you mention, even better! Thank you!

Opened #354 :slightly_smiling_face:

**Comment by dtoakley** [comment]
Hey @pwithnall -- I seem to have a pattern of apologizing for taking a while to get back to you. Sorry for my tardiness. 

You are correct -- it's the "type of network infrastructure the traffic goes over" that would impact the chosen co-efficient. Maybe the user device could affect it too, but it's much more a factor of the network infrastructure (fibre, Wi-Fi, etc). I'm guessing the coefficient could vary a fair bit by the location of the data center (known) and the location of end user (maybe known, but likely not via the cloud provider APIs). 

So yes, we should add the ability to include traffic leaving the data center, we just need to do a bit of research to settle on a co-efficient, or multiple co-efficients for different regions.

**Comment by bpetit** [comment]
Not sure where to put it but I wanted to suggest a little (uncomplete) litterature review, if that helps : https://github.com/cloud-carbon-footprint/cloud-carbon-footprint/issues/379#issuecomment-1291612358
